### ISSUE TITLE
Plan to refactor txn::Scheduler.

### ISSUE THREAD

**Issue by disksing** [pingcap/tikv#807]
Recent benchmark shows that use a threadPool for `txn/scheduler` is not
enough. For example, when all 8 threads in the pool are waiting for Raft engine
to return a snapshot, new requests will be blocked, in other words concurrent
txns may never be more than 8.

We plan to make Engine support async, and refactor Scheduler to increase txn
concurrency.

Engine should supply 2 async interfaces.

```
- async_snapshot(Callback<Snapshot>);
- async_write([Modify], Callback<Result>);
```

Scheduler should split all KV request execution process into 3 stages:
- Stage1: Create snapshot (through Raft).
- Stage2: Execute txn command (on snapshot).
- Stage3: Write modifications (through Raft).

Scheduler is only responsible for scheduling, it maintains ongoing txn's
status, sends Tasks to other threads, when a Task finished, the Scheduler
should decide whether to complete the request or abort it or bring it to
next stage.

We need a thread to do Stage2 txn command execution, or maybe thread pool
if necessary. The thread reads from the snapshot which is produced by Stage1 and
creates modifications which will be written at Stage3.

We also need to maintain a data structure to replace the old `ShardMutex`, which
is used for preventing multiple txns write a same key, typically a
`Map<Key, List<WaitingTxn>>` should be enough, the `Key` maybe hashed to reduce
memory usage.

**Comment by siddontang** [comment]
> Stage2: Execute txn command (local rocksdb).

Using snapshot instead of "local rocksdb" is better?

**Comment by disksing** [comment]
@siddontang Snapshot looks better to me to. The content is updated.

**Comment by siddontang** [comment]
Can we close it now? @disksing

**Comment by disksing** [comment]
Fixed by #850
